hail/src/main/scala/is/hail/utils/FlipbookIterator.scala

```diff
@@ -84,6 +86,65 @@ object FlipbookIterator {
     StagingIterator(sm)
 
   def empty[A] = StagingIterator(StateMachine.terminal[A])
+
+  def multiZipJoin[A: ClassTag](
+    its: IndexedSeq[FlipbookIterator[A]],
+    default: A,
+    ord: (A, A) => Int
+  ): FlipbookIterator[Array[A]] = {
+    val staging = its.map(_.toStagingIterator)
+    val sm = new StateMachine[Array[A]] {
+      private val len = its.length
+      var value: Array[A] = Array.fill(len)(default)
+      var isValid = true
+      var buf = new ArrayBuffer[Int]
```

Comment:
You don't want an `ArrayBuffer`, it is not specialized and will box arguments on every call.  You probably want an `is.hail.utils.ArrayBuilder`.  Note, there a subtle sharing bug in the `ArrayBuilder` in the standard library and it should not be used.